cpmpy/transformations/int2bool.py

```diff
@@ -0,0 +1,469 @@
+        if len(self._xs) <= 1:
+            return []
+        # Encode implication chain `x>=d -> x>=d-1` (using `zip` to create a sliding window)
+        return [curr.implies(prev) for prev, curr in zip(self._xs, self._xs[1:])]
+
+    def _offset(self, d):
+        return d - self._x.lb - 1
+
+    def eq(self, d):
+        """Return a conjunction whether x==d."""
+        if self._x.lb <= d <= self._x.ub:
+            return [self.geq(d), ~self.geq(d + 1)]
+            # return cp.all([self.geq(d), ~self.geq(d + 1)])
+        else:
+            return [BoolVal(False)]
+
+    def geq(self, d):
+        """Return a literal whether x>=d."""
+        if d <= self._x.lb:
+            return BoolVal(True)
+        elif d > self._x.ub:
+            return BoolVal(False)
+        else:
+            return self._xs[self._offset(d)]
+
+    def encode_comparison(self, cmp, d):
+        if cmp == "==":  # x>=d and x<d+1
+            return self.eq(d)
+        elif cmp == "!=":  # x<d or x>=d+1
+            return [cp.any(~lit for lit in self.eq(d))]
+            # TODO we could just use ~cp.all and ~cp.any, but we need to make pysat more robust
+            # return [~self.eq(d)]
+        elif cmp == ">=":
+            return [self.geq(d)]
+        elif cmp == "<=":
+            return [~self.geq(d + 1)]
+        else:
+            raise UNKNOWN_COMPARATOR_ERROR
+
+    def encode_term(self, w=1):
+        return [(w, b) for b in self._xs], w * self._x.lb
+
+
+class IntVarEncLog(IntVarEnc):
+    """
+    Log (or "binary") encoding of an integer variable.
+
+    Uses a Boolean 'bit' variable to represent `x` using the unsigned binary representation offset by its lower bound (e.g. for `x in 5..8`, the assignment `00` maps to `x=5`, and `11` to `x=8`). In other words, it is `k`-offset binary encoding where `k=x.lb`
+    """
+
+    def __init__(self, x):
+        """Create binary encoding of integer variable `x`."""
+        dom_size = _dom_size(x)
+        if dom_size == 0:
+            raise EMPTY_DOMAIN_ERROR
+
+        bits = math.ceil(math.log2(dom_size))
+        if bits == 0:
+            xs = cp.cpm_array([])
+        elif bits == 1:
+            xs = cp.cpm_array([boolvar(name=f"EncBin({x.name})")])
+        else:
+            xs = boolvar(shape=bits, name=f"EncBin({x.name})")
+        super().__init__(x, xs)
+
+    def encode_domain_constraint(self):
+        """Return binary encoding domain constraint (i.e. upper bound is respected with `self._x<=self._x.ub`. The lower bound is automatically enforced by offset binary which maps `000.. = self._x.lb`)."""
+        # encode directly to avoid bounds check for this seemingly tautological constraint
+        return self.encode_comparison("<=", self._x.ub, check_bounds=False)
+
+    def _bitstring(self, d):
+        """Return offset binary representation of `d` as Booleans in order of increasing signicance (e.g. `4` return `001`)."""
```

Comment:
isnt this highly unusual? To have 4 be 001 instead of 100? I find it surprising in any case...

Reply:
you only use this function in 1 place, so it doesn't really matter...

still, this int2bool is supposed to be pedagogical, and its not a regular bitstring that are CS students would expect, but a reversed bitstring, so how about calling the function `_reversed_bitstring(self, d)` and then all is settled? : )